### PR TITLE
Issue #13116: Create separate ant-phase-verify execution for xdocs-examples

### LINKED ISSUE #13116: Create separate ant-phase-verify execution for xdocs-examples
From https://github.com/checkstyle/checkstyle/pull/13112#discussion_r1211651794
> 
> what is benefit to include such files in this validation ?
> https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-non-main-files-checks.xml
> 
> I see no reason to reuse all this validation and fight with suppression later on.
> 
> Lets have one more config and separate execution.

### PR DESCRIPTION
**Pull request by stoyanK7**
Resolves #13116

### CODE REVIEW

**Comment by stoyanK7** [comment]
Right now I am transferring examples and I get the feeling that `// violation` lines should be filtered out from the line length check.

**Comment by nrmancuso** [comment]
> Right now I am transferring examples and I get the feeling that `// violation` lines should be filtered out from the line length check.

I do not agree, especially for documentation, we should limit line length. Do not forget we support `// violation below ...` comments.

File: config/checkstyle-examples-checks.xml
Comment (nrmancuso) [inline config/checkstyle-examples-checks.xml:6]: Please add:
1. File length check (15 lines? 20 lines?)
2. Some check to make sure that top level class name matches `ExampleN`

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Now that we have merged some examples, please change one and show failure of this task in CI. Place a link to the failing build in the PR description.

**Comment by stoyanK7** [comment]
Failing CI with problems I created around.
https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=14739&view=logs&j=eb841e9f-6e1e-5c29-3ad5-74ec2c658069&t=a9b761e6-b8d5-50cf-467a-54967189cab2&l=1274
```
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:1: File does not contain at least 1 matches for pattern '^class Example\d+'. [RegexpSingleline]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Example1.java:1: File length is 33 lines (max allowed is 25). [FileLength]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Example1.java:15:6: File contains tab characters (this is the first instance). [FileTabCharacter]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Example1.java:19: Line is longer than 85 characters (found 103). [LineLength]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Example1.java:19:5: 'method def modifier' has incorrect indentation level 4, expected level should be 2. [Indentation]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Exampleasd.java:1: All example files should start with Example\d+ [RegexpOnFilename]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Exampleasd.java:1: File does not contain at least 1 matches for pattern '^class Example\d+'. [RegexpSingleline]
[ERROR] [checkstyle] [ERROR] /home/vsts/work/1/s/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/Exampleasd.java:1: File length is 32 lines (max allowed is 25). [FileLength]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE

```

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:6]: Done. I put it at 35. If we want less, I have to shorten constantname examples.

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:6]: With the suggestion at https://github.com/checkstyle/checkstyle/pull/13218#issuecomment-1590880632 I would make it even longer for now..

Comment (nrmancuso) [reply config/checkstyle-examples-checks.xml:6]: Good point, please choose number you are comfortable with.

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:6]: Update: We stay with inline, no XML will be added.

I'd stay with 35 personally. 10 lines are lost only for inline comment, package, class declaration, empty line at the end. And this is without defining properties. Sometimes I have to create a method as well just so the code can be compileable. If we add violation comments that need their own line..

Maybe 40-45, do you agree?

Comment (nrmancuso) [reply config/checkstyle-examples-checks.xml:6]: Let’s do 45.

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:6]: Done.

Comment (nrmancuso) [reply config/checkstyle-examples-checks.xml:6]: Also, if we are forbidding `(default)` properties in examples, we should add a module for that, per @romani ’s comment in discord.

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:6]: The problem I have with MatchXPath is that it is `TreeWalker`, wouldn't apply to non-java files.

Comment (nrmancuso) [reply config/checkstyle-examples-checks.xml:6]: Good point :)

File: config/checkstyle-examples-suppressions.xml
Comment (nrmancuso) [inline config/checkstyle-examples-suppressions.xml:10]: If we have issue opened for this, please do `until...` comment. If not, please open one.

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: config/checkstyle-examples-suppressions.xml
Comment (romani) [reply config/checkstyle-examples-suppressions.xml:10]: Yes, I highly recommend to format on creation of examples files. But yeah, suppression is always until issue.

**Comment by romani** [comment]
@stoyanK7 , recheck CI failure

**Comment by stoyanK7** [comment]
@romani All errors are 503 Unavailable from external links(StackOverflow). I don't have control over that
https://app.circleci.com/pipelines/github/checkstyle/checkstyle/18938/workflows/9476f5d1-c401-46e8-b9a8-964aa78d2c21/jobs/308882
```
<td><i><a class="externalLink" href="https://stackoverflow.com/questions/tagged/checkstyle">https://stackoverflow.com/questions/tagged/checkstyle</a>: 503 Service Unavailable</i></td></tr></table></td></tr>
<td><i><a class="externalLink" href="https://stackoverflow.com/questions/tagged/checkstyle">https://stackoverflow.com/questions/tagged/checkstyle</a>: 503 Service Unavailable</i></td></tr></table></td></tr>
<td><i><a class="externalLink" href="https://stackoverflow.com/questions/tagged/checkstyle">https://stackoverflow.com/questions/tagged/checkstyle</a>: 503 Service Unavailable</i></td></tr></table></td></tr>
<td><i><a class="externalLink" href="https://stackoverflow.com/questions/tagged/checkstyle">https://stackoverflow.com/questions/tagged/checkstyle</a>: 503 Service Unavailable</i></td></tr></table></td></tr>
<td><i><a class="externalLink" href="https://stackoverflow.com/questions/tagged/checkstyle">https://stackoverflow.com/questions/tagged/checkstyle</a>: 503 Service Unavailable</i></td></tr></table></td></tr>

```

Opening the link myself it works - https://stackoverflow.com/questions/tagged/checkstyle
No errors when running linkcheck locally

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java
Comment (romani) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: Why ? 
We need this so much.
We need to update check config that complain on such lines

**Review by romani — CHANGES_REQUESTED** [review]
Items

**Review by nrmancuso — CHANGES_REQUESTED** [review]
@stoyanK7 we either need commits to fix violations before introduction of checks, or in same commit. CI should be able to pass on each commit by itself.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java
Comment (nrmancuso) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: You said in discord that we should not include default properties.

**Comment by stoyanK7** [comment]
@nrmancuso Squashing it all in one commit.

**Comment by nrmancuso** [comment]
@stoyanK7 should we add regexp check for presence of example delimiter comments?

**Comment by romani** [comment]
Let's add new checks in separate PRs, we need to quicker merge base validation or most simple, all other enhancements can be un separate PRs.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java
Comment (romani) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: I meant, no defaults in xml format. But in properties format ok to use, as it is not publicly visible.

Comment (stoyanK7) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: @romani Converting configs to `/*xml <module name="..` is part of the plan, right? I don't see a reason to keep the default props.

**Comment by stoyanK7** [comment]
@nrmancuso @romani Added regexp for presence of delimiter comment and added it to examples. Let's finish this because it grows the more other examples get merged.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java
Comment (romani) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: Yes, you are right, it will keep our file size shorter.

Comment (romani) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/constantname/Example1.java:3]: I agree to not having defaults.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java
Comment (romani) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:11]: Please remove xdoc section, we are not ready

**Review by romani — CHANGES_REQUESTED** [review]
Items:

**Comment by romani** [comment]
I highly recommend to not extend this PR with new innovations, lets merge most basic validation sooner.

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java
Comment (stoyanK7) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:11]: Why not? It is less work to add them now while transferring examples to tests?

Comment (romani) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:11]: Hmm, in this case make them to be on its own line

Comment (stoyanK7) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:11]: Makes sense. On it.

**Review by romani — APPROVED** [review]
Ok to merge if CI pass

**Review by nrmancuso — APPROVED** [review]
I am good as CI passes, I just rebased branch after merging other PRs.

**Comment by stoyanK7** [comment]
CI will be red because the PR misses the section comments, I will add them in a minute and push.

**Comment by romani** [comment]
@rnveach , all is ready for final review.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:77]: Convert to `TypeName`.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:63]: Where did 85 come from? I am seeing 100 for our other input files.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:42]: Why not `"^\w+\s*=\s*\(default\).*"`?

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:26]: Why not `^Example\d+\.`?
If so, violation message should be updated too.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:48]: Require to be the only thing on the line.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:56]: Same.

File: config/checkstyle-examples-checks.xml
Comment (rnveach) [inline config/checkstyle-examples-checks.xml:67]: Add `CommentsIndentation`.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: config/checkstyle-examples-checks.xml
Comment (romani) [reply config/checkstyle-examples-checks.xml:63]: We try to enforce some limits that we have now by fact. Just based on some several Example files.
We can improve later to demand shorter files.

Inputs have limit of 120 lines, Example should be shorter.

File: config/checkstyle-examples-checks.xml
Comment (romani) [reply config/checkstyle-examples-checks.xml:67]: Better to add it later on, please create issue to do this in separate PR.
In other case we will never finish first step - just some validation activation.

File: config/checkstyle-examples-checks.xml
Comment (romani) [reply config/checkstyle-examples-checks.xml:63]: Please create issue to not forget to reiterate on exact value later on.

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:63]: Opened at https://github.com/checkstyle/checkstyle/issues/13282

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:67]: Added to https://github.com/checkstyle/checkstyle/issues/13244#issuecomment-1600605964. We are going to add more stuff the moment this PR is merged.

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Done.

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Changed to `^Example\d+\$`. We are ignoring the file extension.

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Done.

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Done.

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:80]: Not sure about this one. `TypeName` is not very flexible in what it's checking. I tried the following
```
    <module name="TypeName">
      <property name="format" value="Example\d+"/>
      <property name="tokens" value="CLASS_DEF"/>
      <property name="applyToPrivate" value="false"/>
      <property name="applyToPublic" value="false"/>
      <property name="applyToProtected" value="false"/>
    </module>
```
```
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:11:18: Name 'AbstractFirst' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:12:18: Name 'Second' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:13:9: Name 'AbstractThird' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:14:9: Name 'Fourth' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:15:18: Name 'GeneratorFifth' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:17:9: Name 'GeneratorSixth' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:12:18: Name 'AbstractFirst' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:13:18: Name 'Second' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example2.java:14:9: Name 'AbstractThird' must match pattern 'Example\d+'. [TypeName]
....
```

https://checkstyle.sourceforge.io/config_naming.html#TypeName

Comment (rnveach) [reply config/checkstyle-examples-checks.xml:80]: @stoyanK7 Please pick one example and provide a link to it why you disagree with violation. Are these sub-classes that are the issue?

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml:80]: Apologies for not having enough info.

Let's take this example. Yes, it has subclasses.
https://github.com/checkstyle/checkstyle/blob/fbbf38a0d9e3517cbea6e41ab95293d6ae20bbb6/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java#L10-L19

```
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:11:18: Name 'AbstractFirst' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:12:18: Name 'Second' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:13:9: Name 'AbstractThird' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:14:9: Name 'Fourth' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:15:18: Name 'GeneratorFifth' must match pattern 'Example\d+'. [TypeName]
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/Example1.java:17:9: Name 'GeneratorSixth' must match pattern 'Example\d+'. [TypeName]

```


I disagree because it cannot be configured to exclude those internal ones. We want to look for a top-level class matching the pattern. If `TypeName` can be configured to do just this, I am fine.

Comment (rnveach) [reply config/checkstyle-examples-checks.xml:80]: Thanks, it looks like this is a limitation of the check. We can ignore for now.

@romani Does it make sense to allow this check to be configured between top level and sub-classes?

File: config/checkstyle-examples-checks.xml
Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: @rnveach 
Would that be okay?
```xml
  <module name="RegexpOnFilename">
    <property name="id" value="exampleFileName"/>
    <property name="fileNamePattern" value="^Example\d+$"/>
    <property name="ignoreFileNameExtensions" value="true"/>
    <property name="match" value="false"/>
    <message key="regexp.filename.mismatch"
             value="All example files should only match '^Example\d+$' exactly"/>
  </module>
```

Comment (rnveach) [reply config/checkstyle-examples-checks.xml]: It is fine, but I still think full regexp here is a little overkill. Also if you are going to do full regexp, I think you can just use `{1}` as they are applied to the message by default somehow.
https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks%2Fregexp+path%3A**%2Fmessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22regexp.filename.mismatch%22&type=code

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Yes, it works.
```diff
$ git diff
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: config/checkstyle-examples-checks.xml
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ config/checkstyle-examples-checks.xml:30 @
    <property name="ignoreFileNameExtensions" value="true"/>
    <property name="match" value="false"/>
    <message key="regexp.filename.mismatch"
-             value="All example files should only match '^Example\d+$' exactly"/>
+             value="All example files should only match ''{1}'' exactly"/>
  </module>

  <module name="RegexpSingleline">
_____________________________________________________________________________________________________________________________________________________________________________
 stoyan @  zenbook:  ~/open-source/checkstyle                                                                                                          issue/13116 [!?↓↑]
$ git status
On branch issue/13116
Your branch and 'origin/issue/13116' have diverged,
and have 6 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   config/checkstyle-examples-checks.xml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/asd.java

no changes added to commit (use "git add" and/or "git commit -a")

```
```
[ERROR] [checkstyle] [ERROR] /home/stoyan/open-source/checkstyle/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/naming/abbreviationaswordinname/asd.java:1: All example files should only match '^Example\d+$' exactly [exampleFileName]
```

Comment (stoyanK7) [reply config/checkstyle-examples-checks.xml]: Done.

File: config/checkstyle-examples-checks.xml
Comment (romani) [reply config/checkstyle-examples-checks.xml:80]: In real life codding, I do not see reason to differentiate rules for top level and nested.

If any need appear, user should use xpath suppression to avoid unwanted violations.
We can use this Check with xpath Suppression. But let's do this in separate issue

Comment (romani) [reply config/checkstyle-examples-checks.xml:80]: Created https://github.com/checkstyle/checkstyle/issues/13290

**Review by rnveach — APPROVED** [review]

**Pull request merged**
